### PR TITLE
no_proxy support

### PR DESCRIPTION
**Pull request by ballwood**
Ported code across from the request library, which I know to work correctly - https://github.com/request/request.

Unsure of where to put the unit/integration tests, if someone could guide me I'll write them. There were 11 tests failing when running npm test when I forked and the same amount after I made the code change.

### CODE REVIEW

**Comment by bnoordhuis** [comment]
Sorry for the delay.  This needs some tests but I'm kind of surprised at how complex it still is.  Is `npm_config_no_proxy` a thing?

**Comment by vbfox** [comment]
I'd really like to see that PR get merged.

For npm config exporting `npm_config_no_proxy` there is a PR open but I don't know if it'll ever get merged: https://github.com/npm/npm/pull/14248 (Currently it can't be specified in npm config)

**Comment by ballwood** [comment]
Hi, sorry I thought I replied to this last week, must have not pressed send. Yeah I noticed that PR was open and included it to future proof it but their work on it seems to have gone by the wayside. 

I can have a look at the tests I'm just trying to understand them at the moment and work out something that works correctly.

**Comment by ballwood** [comment]
@bnoordhuis wrote some small unit tests as I got some free time

File: lib/install.js [hunk 1]
```diff
@@ -423,6 +424,44 @@ function install (gyp, argv, callback) {
 
 }
 
+function formatHostname (hostname) {
+  // canonicalize the hostname, so that 'oogle.com' won't match 'google.com'
+  return hostname.replace(/^\.*/, '.').toLowerCase()
```
Comment (bnoordhuis) [inline lib/install.js]: This transforms `google.com` into `.google.com`.  I infer that is intentional but can you update the comment to make that more explicit?  This behavior also seems only lightly tested.

Aside: can you capitalize and punctuate comments?

File: lib/install.js [hunk 2]
```diff
@@ -423,6 +424,44 @@ function install (gyp, argv, callback) {
...
+}
+
+function parseNoProxyZone (zone) {
+  zone = zone.trim().toLowerCase()
+
+  var zoneParts = zone.split(':', 2)
+  var zoneHost = formatHostname(zoneParts[0])
+  var zonePort = zoneParts[1]
+  var hasPort = zone.indexOf(':') > -1
```
Comment (bnoordhuis) [inline lib/install.js]: Not wrong per se but can also be derived without re-scanning from `zonePort !== undefined` or `zoneParts.length === 2`.

File: lib/install.js [hunk 3]
```diff
@@ -423,6 +424,44 @@ function install (gyp, argv, callback) {
...
+  var port = parsedUri.port || (parsedUri.protocol === 'https:' ? '443' : '80')
+  var hostname = formatHostname(parsedUri.hostname)
+  var noProxyList = noProxy.split(',')
+
+  // iterate through the noProxyList until it finds a match.
+  return noProxyList.map(parseNoProxyZone).some(function (noProxyZone) {
+    var isMatchedAt = hostname.indexOf(noProxyZone.hostname)
+    var hostnameMatched = (
+      isMatchedAt > -1 &&
+      (isMatchedAt === hostname.length - noProxyZone.hostname.length)
```
Comment (bnoordhuis) [inline lib/install.js]: Correct me if I'm wrong but isn't this just a circuitous way of saying `hostname === noProxyZone.hostname`?

File: lib/install.js [hunk 4]
```diff
@@ -438,18 +477,32 @@ function download (gyp, env, url) {
...
-  if (proxyUrl) {
-    if (/^https?:\/\//i.test(proxyUrl)) {
-      log.verbose('download', 'using proxy url: "%s"', proxyUrl)
-      requestOpts.proxy = proxyUrl
-    } else {
-      log.warn('download', 'ignoring invalid "proxy" config setting: "%s"', proxyUrl)
+  var noProxyUrl = gyp.opts.no_proxy
+    || env.no_proxy
+    || env.NO_PROXY
+    || env.npm_config_no_proxy
```
Comment (bnoordhuis) [inline lib/install.js]: This should probably go; the npm pull request wasn't merged.

File: lib/install.js [hunk 5]
```diff
@@ -438,18 +477,32 @@ function download (gyp, env, url) {
...
+    || ''
+
+  if (noProxyUrl === '*' || (noProxyUrl !== '' && uriInNoProxy(url, noProxyUrl))) {
```
Comment (bnoordhuis) [inline lib/install.js]: Long line; can you keep it under 80 columns?  The parens around the `&&` aren't necessary.

File: lib/install.js [hunk 6]
```diff
@@ -438,18 +477,32 @@ function download (gyp, env, url) {
...
+      || env.http_proxy
+      || env.HTTP_PROXY
+      || env.npm_config_proxy
+
+    if (proxyUrl) {
+      if (/^https?:\/\//i.test(proxyUrl)) {
+        log.verbose('download', 'using proxy url: "%s"', proxyUrl)
+        requestOpts.proxy = proxyUrl
+      } else {
+        log.warn('download', 'ignoring invalid "proxy" config setting: "%s"', proxyUrl)
```
Comment (bnoordhuis) [inline lib/install.js]: Ditto, long line.

File: test/test-download-options.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
+var test = require('tape')
```
Comment (bnoordhuis) [inline test/test-download-options.js]: Can you add `'use strict';`?

File: test/test-download-options.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    opts: {
+      proxy: 'http://www.test.com'
+    }
+  }
+
+  var req = install.test.download(gyp, {}, url)
+
+  t.equal(req.opts.proxy, gyp.opts.proxy)
+  t.equal(req.opts.uri, url)
+
```
Comment (bnoordhuis) [inline test/test-download-options.js]: Can you drop the blank line here and below?

**Comment by dlangerenken** [comment]
@bnoordhuis is someone taking care of this PR? Seems like @ballwood is no longer working on it. It would be really nice to have in master.

**Comment by rvagg** [comment]
This looks pretty good and we have `HTTPS_PROXY` support about to land too but it's going to need a new champion. If someone wants this in, please take this code, clean it up as per the suggestions in here and make it work against master. Tests would be good too if you can come up with some that are meaningful.

File: lib/install.js (hunk 3 above)
Comment (alsotang) [reply lib/install.js]: IMO it's a suffix match

**Comment by imatlopez** [comment]
@rvagg I was looking at doing just this given this is quite a pain point inside our corporate network, however is this approach to make it independent of requests, or would it be simpler to set the environment variables from the cli flags as needed and let requests do its job?

**Comment by rvagg** [comment]
@imatlopez you mean `request` not `requests` right? Apparently obeys no_proxy these days but I don't know what we would need to do here, if anything, for that support to flow through. If it's obeying environment variables then I'd assume we don't have to do anything on our end, just set the env var before calling. If you need gyp.opts to be able to carry it then I guess some additional work could go in here and hopefully that'd be simple. You're welcome to open a PR if you'd like it added.

**Pull request closed without merging**